pinot-core/src/main/java/com/linkedin/pinot/core/startree/v2/builder/BaseSingleTreeBuilder.java

```diff
@@ -0,0 +1,477 @@
+   *
+   * @param docId Document Id
+   * @return Star-tree record
+   */
+  abstract Record getStarTreeRecord(int docId) throws IOException;
+
+  /**
+   * Returns the dimension value of the given document and dimension Id in the star-tree.
+   *
+   * @param docId Document Id
+   * @param dimensionId Dimension Id
+   * @return Dimension value
+   */
+  abstract int getDimensionValue(int docId, int dimensionId) throws IOException;
+
+  /**
+   * Sorts and aggregates the records in the segment, and returns a record iterator for all the aggregated records.
+   * <p>This method reads records from segment and generates the initial records for the star-tree.
+   *
+   * @param numDocs Number of documents in the segment
+   * @return Iterator for the aggregated records
+   */
+  abstract Iterator<Record> sortAndAggregateSegmentRecords(int numDocs) throws IOException;
+
+  /**
+   * Generates aggregated records for star-node.
+   * <p>This method will do the following steps:
+   * <ul>
+   *   <li>Creates a temporary buffer for the given range of documents</li>
+   *   <li>Replaces the value for the given dimension Id to {@code STAR}</li>
+   *   <li>Sorts the records inside the temporary buffer</li>
+   *   <li>Aggregates the records with same dimensions</li>
+   *   <li>Returns an iterator for the aggregated records</li>
+   * </ul>
+   *
+   * @param startDocId Start document Id in the star-tree
+   * @param endDocId End document Id (exclusive) in the star-tree
+   * @param dimensionId Dimension Id of the star-node
+   * @return Iterator for the aggregated records
+   */
+  abstract Iterator<Record> generateRecordsForStarNode(int startDocId, int endDocId, int dimensionId)
+      throws IOException;
+
+  /**
+   * Reads the dimensions for a record of the given document Id in the segment.
+   * @param docId
+   * @return
+   */
+  int[] getSegmentRecordDimensions(int docId) {
+    int[] dimensions = new int[_numDimensions];
+    for (int i = 0; i < _numDimensions; i++) {
+      dimensions[i] = _dimensionReaders[i].getDictionaryId(docId);
+    }
+    return dimensions;
+  }
+
+  /**
+   * Reads a record of the given document Id in the segment.
+   *
+   * @param docId Document Id
+   * @return Segment record
+   */
+  Record getSegmentRecord(int docId) {
+    int[] dimensions = getSegmentRecordDimensions(docId);
+    Object[] metrics = new Object[_numMetrics];
+    for (int i = 0; i < _numMetrics; i++) {
+      // Ignore the column for COUNT aggregation function
+      if (_metricReaders[i] != null) {
+        metrics[i] = _metricReaders[i].readSV(docId, _metricDataTypes[i]);
+      }
+    }
+    return new Record(dimensions, metrics);
+  }
+
+  /**
+   * Applies a segment record (raw record) to the current record.
+   *
+   * @param record Current record
+   * @param segmentRecord Segment record
+   * @return Aggregated record
+   */
+  @SuppressWarnings("unchecked")
+  Record applySegmentRecord(@Nullable Record record, Record segmentRecord) {
```

Comment:
Please rename the method and variable names to make it more readable. For example s/apply/merge, s/record/mergedRecord, etc

Reply:
Done